File: testdata/jennies/rawtypes/struct_with_complex_fields/JavaRawTypes/struct_complex_fields/SomeStruct.java
package struct_complex_fields;

import java.util.Objects;
import java.util.LinkedList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// This struct does things.
public class SomeStruct {
    public SomeOtherStruct fieldRef;
    public StringOrBool fieldDisjunctionOfScalars;
    public StringOrSomeOtherStruct fieldMixedDisjunction;
    public String fieldDisjunctionWithNull;
    public SomeStructOperator operator;
    public List<String> fieldArrayOfStrings;
    public Map<String, String> fieldMapOfStringToString;
    public StructComplexFieldsSomeStructFieldAnonymousStruct fieldAnonymousStruct;
    public String fieldRefToConstant;
    public SomeStruct() {
        this.fieldRef = new struct_complex_fields.SomeOtherStruct();
        this.fieldDisjunctionOfScalars = new struct_complex_fields.StringOrBool();
        this.fieldMixedDisjunction = new struct_complex_fields.StringOrSomeOtherStruct();
        this.operator = SomeStructOperator.GREATER_THAN;
        this.fieldArrayOfStrings = new LinkedList<>();
        this.fieldMapOfStringToString = new HashMap<>();
        this.fieldAnonymousStruct = new struct_complex_fields.StructComplexFieldsSomeStructFieldAnonymousStruct();
        this.fieldRefToConstant = new struct_complex_fields.ConnectionPath();
    }
    public SomeStruct(SomeOtherStruct fieldRef,StringOrBool fieldDisjunctionOfScalars,StringOrSomeOtherStruct fieldMixedDisjunction,String fieldDisjunctionWithNull,SomeStructOperator operator,List<String> fieldArrayOfStrings,Map<String, String> fieldMapOfStringToString,StructComplexFieldsSomeStructFieldAnonymousStruct fieldAnonymousStruct,String fieldRefToConstant) {
        this.fieldRef = fieldRef;
        this.fieldDisjunctionOfScalars = fieldDisjunctionOfScalars;
        this.fieldMixedDisjunction = fieldMixedDisjunction;
        this.fieldDisjunctionWithNull = fieldDisjunctionWithNull;
        this.operator = operator;
        this.fieldArrayOfStrings = fieldArrayOfStrings;
        this.fieldMapOfStringToString = fieldMapOfStringToString;
        this.fieldAnonymousStruct = fieldAnonymousStruct;
        this.fieldRefToConstant = fieldRefToConstant;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        if (!(other instanceof SomeStruct)) return false;
        SomeStruct o = (SomeStruct) other;
        if (!Objects.equals(this.fieldRef, o.fieldRef)) return false;
        if (!Objects.equals(this.fieldDisjunctionOfScalars, o.fieldDisjunctionOfScalars)) return false;
        if (!Objects.equals(this.fieldMixedDisjunction, o.fieldMixedDisjunction)) return false;
        if (!Objects.equals(this.fieldDisjunctionWithNull, o.fieldDisjunctionWithNull)) return false;
        if (!Objects.equals(this.operator, o.operator)) return false;
        if (!Objects.equals(this.fieldArrayOfStrings, o.fieldArrayOfStrings)) return false;
        if (!Objects.equals(this.fieldMapOfStringToString, o.fieldMapOfStringToString)) return false;
        if (!Objects.equals(this.fieldAnonymousStruct, o.fieldAnonymousStruct)) return false;
        if (!Objects.equals(this.fieldRefToConstant, o.fieldRefToConstant)) return false;
        return true;
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.fieldRef, this.fieldDisjunctionOfScalars, this.fieldMixedDisjunction, this.fieldDisjunctionWithNull, this.operator, this.fieldArrayOfStrings, this.fieldMapOfStringToString, this.fieldAnonymousStruct, this.fieldRefToConstant);
    }
}

File: testdata/jennies/rawtypes/struct_with_complex_fields/PHPRawTypes/src/StructComplexFields/SomeStruct.php
<?php

namespace Grafana\Foundation\StructComplexFields;

/**
 * This struct does things.
 */
class SomeStruct implements \JsonSerializable
{
    public \Grafana\Foundation\StructComplexFields\SomeOtherStruct $fieldRef;

    public \Grafana\Foundation\StructComplexFields\StringOrBool $fieldDisjunctionOfScalars;

    public \Grafana\Foundation\StructComplexFields\StringOrSomeOtherStruct $fieldMixedDisjunction;

    public ?string $fieldDisjunctionWithNull;

    public \Grafana\Foundation\StructComplexFields\SomeStructOperator $operator;

    /**
     * @var array<string>
     */
    public array $fieldArrayOfStrings;

    /**
     * @var array<string, string>
     */
    public array $fieldMapOfStringToString;

    public \Grafana\Foundation\StructComplexFields\StructComplexFieldsSomeStructFieldAnonymousStruct $fieldAnonymousStruct;

    public string $fieldRefToConstant;

    /**
     * @param \Grafana\Foundation\StructComplexFields\SomeOtherStruct|null $fieldRef
     * @param \Grafana\Foundation\StructComplexFields\StringOrBool|null $fieldDisjunctionOfScalars
     * @param \Grafana\Foundation\StructComplexFields\StringOrSomeOtherStruct|null $fieldMixedDisjunction
     * @param string|null $fieldDisjunctionWithNull
     * @param \Grafana\Foundation\StructComplexFields\SomeStructOperator|null $operator
     * @param array<string>|null $fieldArrayOfStrings
     * @param array<string, string>|null $fieldMapOfStringToString
     * @param \Grafana\Foundation\StructComplexFields\StructComplexFieldsSomeStructFieldAnonymousStruct|null $fieldAnonymousStruct
     * @param \Grafana\Foundation\StructComplexFields\ConnectionPath|null $fieldRefToConstant
     */
    public function __construct(?\Grafana\Foundation\StructComplexFields\SomeOtherStruct $fieldRef = null, ?\Grafana\Foundation\StructComplexFields\StringOrBool $fieldDisjunctionOfScalars = null, ?\Grafana\Foundation\StructComplexFields\StringOrSomeOtherStruct $fieldMixedDisjunction = null, ?string $fieldDisjunctionWithNull = null, ?\Grafana\Foundation\StructComplexFields\SomeStructOperator $operator = null, ?array $fieldArrayOfStrings = null, ?array $fieldMapOfStringToString = null, ?\Grafana\Foundation\StructComplexFields\StructComplexFieldsSomeStructFieldAnonymousStruct $fieldAnonymousStruct = null, ?\Grafana\Foundation\StructComplexFields\ConnectionPath $fieldRefToConstant = null)
    {
        $this->fieldRef = $fieldRef ?: new \Grafana\Foundation\StructComplexFields\SomeOtherStruct();
        $this->fieldDisjunctionOfScalars = $fieldDisjunctionOfScalars ?: new \Grafana\Foundation\StructComplexFields\StringOrBool();
        $this->fieldMixedDisjunction = $fieldMixedDisjunction ?: new \Grafana\Foundation\StructComplexFields\StringOrSomeOtherStruct();
        $this->fieldDisjunctionWithNull = $fieldDisjunctionWithNull;
        $this->operator = $operator ?: \Grafana\Foundation\StructComplexFields\SomeStructOperator::GreaterThan();
        $this->fieldArrayOfStrings = $fieldArrayOfStrings ?: [];
        $this->fieldMapOfStringToString = $fieldMapOfStringToString ?: [];
        $this->fieldAnonymousStruct = $fieldAnonymousStruct ?: new \Grafana\Foundation\StructComplexFields\StructComplexFieldsSomeStructFieldAnonymousStruct();
        $this->fieldRefToConstant = $fieldRefToConstant ?: \Grafana\Foundation\StructComplexFields\ConnectionPath;
    }

    /**
     * @param array<string, mixed> $inputData
     */
    public static function fromArray(array $inputData): self
    {
        /** @var array{FieldRef?: mixed, FieldDisjunctionOfScalars?: mixed, FieldMixedDisjunction?: mixed, FieldDisjunctionWithNull?: string, Operator?: string, FieldArrayOfStrings?: array<string>, FieldMapOfStringToString?: array<string, string>, FieldAnonymousStruct?: mixed, fieldRefToConstant?: string} $inputData */
        $data = $inputData;
        return new self(
            fieldRef: isset($data["FieldRef"]) ? (function($input) {
    	/** @var array{FieldAny?: mixed} */
    $val = $input;
    	return \Grafana\Foundation\StructComplexFields\SomeOtherStruct::fromArray($val);
    })($data["FieldRef"]) : null,
            fieldDisjunctionOfScalars: isset($data["FieldDisjunctionOfScalars"]) ? (function($input) {
    	/** @var array{String?: string, Bool?: bool} */
    $val = $input;
    	return \Grafana\Foundation\StructComplexFields\StringOrBool::fromArray($val);
    })($data["FieldDisjunctionOfScalars"]) : null,
            fieldMixedDisjunction: isset($data["FieldMixedDisjunction"]) ? (function($input) {
    	/** @var array{String?: string, SomeOtherStruct?: mixed} */
    $val = $input;
    	return \Grafana\Foundation\StructComplexFields\StringOrSomeOtherStruct::fromArray($val);
    })($data["FieldMixedDisjunction"]) : null,
            fieldDisjunctionWithNull: $data["FieldDisjunctionWithNull"] ?? null,
            operator: isset($data["Operator"]) ? (function($input) { return \Grafana\Foundation\StructComplexFields\SomeStructOperator::fromValue($input); })($data["Operator"]) : null,
            fieldArrayOfStrings: $data["FieldArrayOfStrings"] ?? null,
            fieldMapOfStringToString: $data["FieldMapOfStringToString"] ?? null,
            fieldAnonymousStruct: isset($data["FieldAnonymousStruct"]) ? (function($input) {
    	/** @var array{FieldAny?: mixed} */
    $val = $input;
    	return \Grafana\Foundation\StructComplexFields\StructComplexFieldsSomeStructFieldAnonymousStruct::fromArray($val);
    })($data["FieldAnonymousStruct"]) : null,
            fieldRefToConstant: isset($data["fieldRefToConstant"]) ? /* ref to a non-struct, non-enum, this should have been inlined */ (function(array $input) { return $input; })($data["fieldRefToConstant"]) : null,
        );
    }

    /**
     * @return mixed
     */
    public function jsonSerialize(): mixed
    {
        $data = new \stdClass;
        $data->FieldRef = $this->fieldRef;
        $data->FieldDisjunctionOfScalars = $this->fieldDisjunctionOfScalars;
        $data->FieldMixedDisjunction = $this->fieldMixedDisjunction;
        $data->Operator = $this->operator;
        $data->FieldArrayOfStrings = $this->fieldArrayOfStrings;
        $data->FieldMapOfStringToString = $this->fieldMapOfStringToString;
        $data->FieldAnonymousStruct = $this->fieldAnonymousStruct;
        $data->fieldRefToConstant = $this->fieldRefToConstant;
        if (isset($this->fieldDisjunctionWithNull)) {
            $data->FieldDisjunctionWithNull = $this->fieldDisjunctionWithNull;
        }
        return $data;
    }
}
